### ISSUE TITLE
Bug: no-unassigned-vars should ignore declarations with a TS definite assignment assertion

### ISSUE THREAD

**Issue by kirkwaiblinger** [eslint/eslint#20051]
### Environment

playground

### What parser are you using?

@typescript-eslint/parser

### What did you do?

```ts
/* eslint no-unassigned-vars: ["warn"] */

let x!: string;

eval("x = 'a string'");

x.toString();
```

### What did you expect to happen?

no report

### What actually happened?

lint report: 'x' is always 'undefined' because it's never assigned.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVuYXNzaWduZWQtdmFyczogW1wid2FyblwiXSAqL1xuXG5sZXQgeCE6IHN0cmluZztcblxuZXZhbChcInggPSAnYSBzdHJpbmcnXCIpO1xuXG54LnRvU3RyaW5nKCk7Iiwib3B0aW9ucyI6eyJydWxlcyI6e30sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9LCJzb3VyY2VUeXBlIjoibW9kdWxlIn0sInBhcnNlciI6IkB0eXBlc2NyaXB0LWVzbGludC9wYXJzZXIifX19

### Participation

- [x] I am willing to submit a pull request for this issue.

### Additional comments

IMO this shouldn't report since the definite assignment assertion tells TS "I promise you this is assigned by the team you read it"

Observed in the wild: https://github.com/typescript-eslint/typescript-eslint/blob/05014ff054cfa336c8ca48bac08deec3225675ab/packages/eslint-plugin/tools/generate-breaking-changes.mts#L22-L27

**Comment by kirkwaiblinger** [comment]
I've thought about this a little more and I think it's actually probably a bad idea. The (direct) `eval` edge case is extremely niche, and easy enough to suppress with a disable comment.

The standard use of definite assignment assertions is to do something like the following, to suppress use-before-define errors on the return assignment. 


```ts
function doSomething(): string {
    let x!: string;

    function assignX() {
        x = 'a string';
    }

    assignX();

    return x;
}
```

In this case, it would be harmful to disable no-unassigned-vars since it could catch provably unsound uses of the definite assignment assertion

```ts
let y: string = 'something';

function doSomething(): string {
    let x!: string;

    function assignX() {
        y = 'a string'; // typo
    }

    assignX();

    return x; // provably returns without initializing x to a string
}
```

So, I think my proposal is a bad idea actually.
